File: Introductory/DigitQueries.cpp
#include <bits/stdc++.h>
#define int long long
using namespace std;
char solve(int query) {
    if(query<=9) return '0'+query;
    int totalDigitsTillNow = 9;
    int individualNumberOfDigits = 1;
    int numbersWithCurrentIndividualDigitCount = 9;
    while(totalDigitsTillNow<query) {
        individualNumberOfDigits++;
        numbersWithCurrentIndividualDigitCount*=10;
        totalDigitsTillNow+=(individualNumberOfDigits*numbersWithCurrentIndividualDigitCount);
    }
    totalDigitsTillNow-=(individualNumberOfDigits*numbersWithCurrentIndividualDigitCount);
    int startingNumber = pow(10, individualNumberOfDigits-1);
    int blockQueryNumber = query-totalDigitsTillNow;
    int whatToAdd = blockQueryNumber/individualNumberOfDigits;
    if(blockQueryNumber%individualNumberOfDigits==0) whatToAdd--;
    int currentNumber = startingNumber+whatToAdd;
    int whatDigit = blockQueryNumber%individualNumberOfDigits;
    if(whatDigit==0) whatDigit = individualNumberOfDigits;
    whatDigit--;
    return to_string(currentNumber)[whatDigit];
}
int32_t main() {
    #ifndef ONLINE_JUDGE
    freopen("/Users/keshavpaliwal/Desktop/Everything/CP/input.txt", "r", stdin);
    freopen("/Users/keshavpaliwal/Desktop/Everything/CP/output.txt", "w", stdout);
    #endif
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int t; cin >> t;
    while(t--) {
        int q; cin >> q;
        cout << solve(q) << '\n';
    }
    return 0;
}

File: Introductory/DigitQueries.py
q = int(input())
for _ in range(q):
    k = int(input())
    numberOfDigits = 1
    while 9 * numberOfDigits * pow(10, numberOfDigits - 1) < k:
        k -= (9 * numberOfDigits * pow(10, numberOfDigits - 1))
        numberOfDigits += 1
    currentNumber = pow(10, numberOfDigits - 1) + (
        (k // numberOfDigits) - 1 if k % numberOfDigits == 0 else k // numberOfDigits)
    k %= numberOfDigits
    if k == 0:
        k = numberOfDigits
    k -= 1
    print(str(currentNumber)[k])
